spec/21-http_header_format_rationale.md

```diff
@@ -16,6 +16,103 @@ Making `trace-flags` optional doesn't save a lot, but makes specification more c
+may use only a subset of `trace-id` bytes to read and set `trace-id`. This
+behavior will break interoperability with vendors and platforms operating with
+the longer identifiers. If only a subset of `trace-id` bytes were read on
+incoming requests and sent with the outgoing call, for the systems operating
+with longer identifiers incoming and outgoing `trace-id` will not match. These
+systems will identify this situation as a restarted trace.
+
+Note, that vendors and platforms may implement a special logic to interoperate
+with the systems like this.
+
+Specification uses "SHOULD language" when asking to fill up extra bytes with
+random numbers instead of zeros. Typically, tracing systems that will not
+propagate extra bytes of incoming `trace-id` will not follow this ask and will
+fill up extra bytes with zeros.
+
+#### Linking to the longer thread-id
+
+As a minor step to improve interoperability between tracing systems, system that
+operates with shorter identifiers may record a longer incoming `trace-id` as a
+property of a telemetry item representing the incoming request.
+
+#### Propagating extra bytes of trace-id
+
+Preserving the `trace-id` unchanged is a major improvement in interoperability
+of tracing systems using different number of `trace-id` bytes. It is typical
+that tracing systems may propagate extra information from incoming request to
+outgoing calls. While recording of these extra bytes to the tracing systems
+backend is not possible.
+
+If tracing system can propagate these extra bytes, than it MUST do it.
+
+When tracing system has this capability, specification suggests that extra bytes
+are fill out with random numbers instead of zeros. This requirement helps to
+validate that tracing systems implements propagation from incoming request to
+outgoing calls correctly.
+
+#### Propagating tracestate as well
+
+Note, that even greater interoperability will be achieved with propagating a
+`tracestate` header. Dropping this header may break vendor specific distributed
+tracing scenarios. But this behavior conforms to the specification and older
+tracing systems may do it.
+
+As noted in previously, these tracing systems must do a best effort of
+propagating this header even if it will not be recorded to the tracing system
+backend.
+
+## Trace ID randomization "left padding"
+
+Specification explains the "left padding" requirement for the random `trace-id`
+generation. Tracing systems will implement various algorithms that use
+`trace-id` as a base to make a sampling decision. Typically, it will be some
+variation of hash calculation. Those algorithms may be giving different "weight"
+to different bytes of a `trace-id`. So the requirement to keep randomness to the
+left helps interoperability between tracing systems by suggesting which bytes
+carry a bigger weight in hash calculation.
+
+Practically, there will be tracing systems filling first bytes with zeros (see
```

Comment:
I would much rather see the specification reverse expectations here. IMO systems `SHOULD` backfill with zeros, and there could be a note that since it's only a `SHOULD` and not a requirement, then systems could choose to do the random 64 bit backfill if they feel they need to due to a hash-based-left-bytes sampling algorithm.

In other words, default to maximum interoperability, and point out where there might be valid use cases for doing something else. As it stands, this feels backwards.

The reality is that implementors will see `SHOULD`, and the language around 64 bit trace IDs being "violations", and they will blindly backfill with random 64 bits. As a result we'll see breakage by default instead of interoperability by default.

Reply:
I still don't follow what system will break. And why the randomness requirement will break it? It felt almost like you suggesting all systems only fill up 64 bits.

In the doc there are three types of systems described. First, which simply ignores extra bytes. These systems definitely will be better off by filling up with zeros. And they can. They already breaking couple MUST of this spec. Ignoring second. Third, that only recording 64 but respecting others and propagates 128. For this tracing systems - why not ask to fill up extra bytes with random numbers? What difference does it make? It still will do it when some external 128bit system will call into it and extra bytes are needed to be propagated.

Reply:
> why not ask to fill up extra bytes with random numbers?

Try looking at this from the other direction - why not fill up with zeros? Multiple people are telling you that backfilling with zeros will allow for greater compatibility with 64 bit systems.